### ISSUE TITLE
IRBuilder for eo

### ISSUE THREAD

**Issue by levBagryansky** [objectionary/eo#2813]
Right now, all we have to modify `xmir` is `xsl` and java tools for working with `xml.` In my opinion, this is not the most convenient and appropriate way to work with the intermediate representation of the language, since it forces you to work with the language at the level and in terms of `xml`: `Node`s, `attribute`s, and so on. Perhaps `xml` does a good job of representing a program in `eo`, but working with it is not abstract and high-level enough. For example, try changing the name of an object that is used in many places. Or try adding another abstract object as an attribute to the object - at the xmir level, this is not a trivial task.
As for me, an example of a more convenient interaction with an internal representation is `LLVM`. They do not change the file itself, but interact with a special api, where there is, for example, the `IRBuilder` class- to add instructions or functions, `BasicBlock`, `Instruction`, and so on. 
Moving away from the topic, they have a rather complicated class hierarchy for all occasions (https://llvm.org/doxygen/classllvm_1_1Value.html ). 
In short, they work in terms of the entities of their ir, not in terms of a specific representation. 
Therefore, it seems to me it would be convenient to make an abstract api specifically for working with `eo` and in terms of `eo` from the outside, and rely on `xmir` inside.

**Comment by yegor256** [comment]
@levBagryansky good idea! What would be the interface of this API? Can you draft an example: how one can build/modify a EO program using API calls?

**Comment by c71n93** [comment]
There is simple example of analyzing EO code with IRBuilder, that may be useful.
```java
IRBuilder builder = new XMIRBuilder("example.xmir");
AbstractObjects abstracts = builder.getAbstracts();
for (abstr : abstracts) {
    if (abstr.isDecorator()) {
        System.out.println(abstr.name() + " decorates " + abstr.getAttr("@").base().toString())
    }
}
```
This code should for every top level abstract object print what it decorates. 

For this EO example:
```
[] > parent
  [] > attr1
    1 > @

[] > child
  parent > @
  [] > attr2
    2 > @
```
Output will be:
```
child decorates parent
```
@levBagryansky Do you have any thoughts?

**Comment by levBagryansky** [comment]
```java
IRBuilder builder = new XMIRBuilder("example.xmir");
Collection<AbstractObj> abstracts = builder.getAbstracts();
for (AbstractObj abstr : abstracts) {
    if (abstr.hasAttr("@")) {
        System.out.println(abstr.name() + " decorates " + abstr.getAttr("@").base().toString())
    }
}
```

**Comment by levBagryansky** [comment]
@c71n93 take a look at the eo code:
```
[] > child
  parent.attr1 > @
  [] > attr2
    2 > @
```
We can split this `NamedObject` into two parts: right part is the name `"child"` and left part is an expression:
```
[]
  parent.attr1 > @
  [] > attr2
    2 > @
```
This like "base" in XMIR.
1) Expression can be an `abstraction` like in the example above. Then it has a list of `attribute`s (free and bound). We have taken the induction step because each attribute is similar to the example we are considering at the beginning.
2) Expression an application:
`2.plus 3 4 5`
There is an `abstraction` with following parts: source- `2.plus`, and args: `3`, `4`, `5`.
3) Expression can be  a `dot.notation`: `2.plus` is a `dot.notation`.
4) Expression can be a `link` to other object.
Thus the API can be following:
```java
List<NamedObject> objs = builder.getNamed();
NamedObject child = objs.stream().filter(obj -> "child".equal(obj.name())).getFirst();
sout(child.name()); // "child"
Expression abstr = child.value(); 
sout(abstr.getClass());  // Abstraction
NamedObj firstAttr = ((Abstraction) abstr).attr(0 /* or "@" */); // parent.attr1 > @
sout("should be @: " + firstAttr).name())
firstAttr.value()   // Expression that is DotNotation, parent.attr1
    .getArg(0)        // parent
    .getClass()       // Link
```

**Comment by levBagryansky** [comment]
@yegor256 please take a look

**Comment by yegor256** [comment]
@levBagryansky I suggest we do this in a separate repository
